samples/net/http_server/src/main.c

```diff
@@ -4,144 +4,294 @@
+#else
+#define SYS_LOG_DOMAIN "http-server"
+#endif
+#define NET_SYS_LOG_LEVEL SYS_LOG_LEVEL_DEBUG
+#define NET_LOG_ENABLED 1
+#endif
+
 #include <zephyr.h>
-#include <net/net_context.h>
+#include <stdio.h>
 
-#include <misc/printk.h>
+#include <net/net_context.h>
+#include <net/http.h>
 
-#if defined(CONFIG_NET_L2_BLUETOOTH)
-#include <bluetooth/bluetooth.h>
-#include <gatt/ipss.h>
-#endif
+#include <net_sample_app.h>
 
-#include "http_types.h"
-#include "http_server.h"
-#include "http_utils.h"
-#include "http_write_utils.h"
 #include "config.h"
 
 /* Sets the network parameters */
-static
-int network_setup(struct net_context **net_ctx, net_tcp_accept_cb_t accept_cb,
-		  const char *addr, u16_t port);
 
-#if defined(CONFIG_MBEDTLS)
-#include "ssl_utils.h"
-#endif
+#define RESULT_BUF_SIZE 1024
+static u8_t http_result[RESULT_BUF_SIZE];
 
-void main(void)
+#if defined(CONFIG_HTTPS)
+#if !defined(CONFIG_HTTPS_STACK_SIZE)
+#define CONFIG_HTTPS_STACK_SIZE		8192
+#endif /* CONFIG_HTTPS_STACK_SIZE */
+
+#define APP_BANNER "Run HTTPS server"
+#define INSTANCE_INFO "Zephyr HTTPS example server #1"
+
+/* Note that each HTTPS context needs its own stack as there will be
+ * a separate thread for each HTTPS context.
+ */
+NET_STACK_DEFINE(HTTPS, https_stack, CONFIG_HTTPS_STACK_SIZE,
+		 CONFIG_HTTPS_STACK_SIZE);
+
+#define RX_FIFO_DEPTH 4
+K_MEM_POOL_DEFINE(ssl_rx_pool, 4, 64, RX_FIFO_DEPTH, 4);
+
+static struct http_server_ctx https_ctx;
+
+static u8_t https_result[RESULT_BUF_SIZE];
+
+#else /* CONFIG_HTTPS */
+#define APP_BANNER "Run HTTP server"
+#endif /* CONFIG_HTTPS */
+
+/*
+ * Note that the http_server_ctx and http_server_urls are quite large so be
+ * careful if those are allocated from stack.
+ */
+static struct http_server_ctx http_ctx;
+static struct http_server_urls http_urls;
+
+void panic(const char *msg)
 {
-#if defined(CONFIG_NET_L2_BLUETOOTH)
-	int err;
+	if (msg) {
+		NET_ERR("%s", msg);
+	}
+
+	for (;;) {
+		k_sleep(K_FOREVER);
+	}
+}
+
+#define HTTP_STATUS_200_OK	"HTTP/1.1 200 OK\r\n" \
+				"Content-Type: text/html\r\n" \
+				"Transfer-Encoding: chunked\r\n" \
+				"\r\n"
 
-	err = bt_enable(NULL);
-	if (err) {
-		printk("Bluetooth init failed (err %d)\n", err);
+#define HTML_HEADER		"<html><head>" \
+				"<title>Zephyr HTTP Server</title>" \
+				"</head><body><h1>" \
+				"<center>Zephyr HTTP server</center></h1>\r\n"
+
+#define HTML_FOOTER		"</body></html>\r\n"
+
+/* Prints the received HTTP header fields as an HTML list */
+static void print_http_headers(struct http_server_ctx *ctx,
+			       char *str, u16_t size)
+{
+	u16_t offset = 0;
+
+	offset = snprintf(str, size,
+			  HTML_HEADER
+			  "<h2>HTTP Header Fields</h2>\r\n<ul>\r\n");
+	if (offset >= size) {
 		return;
 	}
 
-	printk("Bluetooth initialized\n");
+	for (u8_t i = 0; i < ctx->req.field_values_ctr; i++) {
+		struct http_field_value *kv = &ctx->req.field_values[i];
 
-	ipss_init();
+		offset += snprintf(str + offset, size - offset,
```

Comment:
`snprintf()` returns an integer, and you're accumulating its return value on an unsigned short. In the minimal libc, it can return EOF, which is defined to -1; I don't know about newlib.  It's better to store the return value on an integer, check for overflow (`if (r < 0 || r >= size_param)`), and only then accumulate to the offset variable -- making sure it's not going to go past `MAX_UINT16`.